packages/core-app-api/src/apis/implementations/IdentityApi/startCookieAuthRefresh.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { DiscoveryApi, ErrorApi, FetchApi } from '@backstage/core-plugin-api';
+
+const PLUGIN_ID = 'app';
+const CHANNEL_ID = `${PLUGIN_ID}-auth-cookie-expires-at`;
+
+const MIN_BASE_DELAY_MS = 5 * 60_000;
+
+const ERROR_BACKOFF_START = 5_000;
+const ERROR_BACKOFF_FACTOR = 2;
+const ERROR_BACKOFF_MAX = 5 * 60_000;
+
+// Messaging implementation and IDs must match
+//   plugins/auth-react/src/hooks/useCookieAuthRefresh/useCookieAuthRefresh.tsx
+export function startCookieAuthRefresh({
+  discoveryApi,
+  fetchApi,
+  errorApi,
+}: {
+  discoveryApi: DiscoveryApi;
+  fetchApi: FetchApi;
+  errorApi: ErrorApi;
+}) {
+  let stopped = false;
+  let timeout: NodeJS.Timeout | undefined;
+  let firstError = true;
+  let errorBackoff = ERROR_BACKOFF_START;
+
+  const channel =
+    'BroadcastChannel' in window ? new BroadcastChannel(CHANNEL_ID) : undefined;
+
+  // Randomize the refreshing margin with a margin of 1-4 minutes to avoid all tabs refreshing at the same time
+  const getDelay = (expiresAt: number) => {
+    const margin = (1 + 3 * Math.random()) * 60000;
+    const delay = Math.max(expiresAt - Date.now(), MIN_BASE_DELAY_MS) - margin;
+    return delay;
+  };
+
+  const refresh = async () => {
+    try {
+      const baseUrl = await discoveryApi.getBaseUrl(PLUGIN_ID);
+      const requestUrl = `${baseUrl}/.backstage/auth/v1/cookie`;
+      const res = await fetchApi.fetch(`${requestUrl}`, {
+        credentials: 'include',
+      });
+
+      if (!res.ok) {
+        throw new Error(
+          `Request failed with status ${res.status} ${res.statusText}, see request towards ${requestUrl} for more details`,
+        );
+      }
+
+      firstError = true;
```

Comment:
Hm I suppose you put it here to kinda distinguish between classes of errors, but not sure it's helpful 🤔 could go all the way at the bottom before the catch.

For example if you accidentally introduce some json error in code, it'll never log, otherwise, because it'll constantly look like it's the first error

Reply:
Ah no it just remained here after a refactor actually, moved it down